### PR TITLE
Create API to search by parcel and block

### PR DESCRIPTION
**Pull request by harshitsan**

### CODE REVIEW

File: server/api/apiRoutes.js [hunk 1]
```diff
@@ -1,81 +1,84 @@
-const Router = require('express').Router();
```
Comment (GalGend) [inline server/api/apiRoutes.js:1]: Can you please remove the linting changes?

File: server/api/controller/shape.js [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+                    data: result[0],
+                });
+            }
+            res.status(200).json({ status: 'OK', data: result[0] });
+        })
+        .catch((err) => {
+            req.error = err;
+            next();
+        });
+};
```
Comment (GalGend) [inline server/api/controller/shape.js:82]: Can you please remove the package.lock?

File: client/src/pages/Tree/common/Views/shareTree.jsx [hunk 3]
```diff
@@ -108,7 +108,7 @@ const CopyUrlArea = withTheme(styled.div`
 const ShareTreeView = () => {
 	const theme = useTheme();
 	const url = encodeURI(window.location.toString());
-	const whatsappURL = `https://wa.me/?text=רשיון%20כריתת%20עצים%20שאולי%20יעניין%20אותך%3A%0A${url}`;
+	const whatsappURL = `https://wa.me/?text=היתר%20שאולי%יעניין%20אותך%3A%0A${url}`;
```
Comment (yaelal) [inline client/src/pages/Tree/common/Views/shareTree.jsx:111]: this should not be changed

File: client/src/components/AutoCompleteParcleInput.jsx [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+            {...itemProps}
+            key={suggestion.id}
+            selected={isHighlighted}
+            component="div"
+            style={{
+                fontWeight: isSelected ? 500 : 400,
+                fontFamily: 'Assistant',
+            }}
+        >
+            {/* {console.log(suggestion)} */}
```
Comment (GalGend) [inline client/src/components/AutoCompleteParcleInput.jsx:74]: please remove comments

**Review by gruppin — CHANGES_REQUESTED** [review]
hi @harshitsan! please revert all the linter-only edits and keep only the real changes. thanks

**Review by gruppin — CHANGES_REQUESTED** [review]
hi @harshitsan!  
thanks for your effort!
please review and address the issues inline.

File: client/src/components/AutoCompleteInput.jsx [hunk 5]
```diff
@@ -10,167 +10,188 @@ import Downshift from 'downshift';
...
+    );
 }
 
 function renderSuggestion({
-  suggestion,
-  index,
-  itemProps,
-  highlightedIndex,
-  selectedItem,
+    suggestion,
```
Comment (GalGend) [inline client/src/components/AutoCompleteInput.jsx:53]: Fix linting here, unacceptable

File: client/src/components/Autocomplete/Autocomplete.jsx [hunk 6]
```diff
@@ -5,71 +5,81 @@ import MenuItem from '@material-ui/core/MenuItem';
...
+    selected: {
+        backgroundColor: `${colors.grey[300]} !important`,
+    },
+})(MenuItem);
 
 function Autocomplete(props) {
-  const { placeholder, width, icon, items, onInputChange, onFilter } = props;
-  const [inputValue, setInputValue] = useState('');
-  const [selectedItem, setSelectedItem] = useState([]);
+    const { placeholder, width, icon, items, onInputChange, onFilter } = props;
```
Comment (GalGend) [inline client/src/components/Autocomplete/Autocomplete.jsx:19]: Fix linting, unacceptable

File: client/src/components/Mapa.jsx [hunk 7]
```diff
@@ -3,63 +3,88 @@ import leaflet from 'leaflet';
...
+        geom,
+        countyName,
+        maxZoom = 17,
+        showPlaceholder = false,
+    } = props;
 
-const Mapa = (props) =>  {
-	const { hideZoom, disableInteractions, title2, geom, countyName, maxZoom=17, showPlaceholder=false } = props;
+    if (!geom || geom.length === 0) {
+        if (showPlaceholder) {
```
Comment (GalGend) [inline client/src/components/Mapa.jsx:18]: Fix linting, unacceptable

File: client/src/redux/reducers.js [hunk 8]
```diff
@@ -7,13 +7,12 @@ import fundingReducer from './funding/slice';
 import treeReducer from './tree/slice';
 
 const rootReducer = combineReducers({
-	modal: modalReducer,
-	user: userReducer,
-	comments: commentsReducer,
-	plan: planReducer,
-	funding: fundingReducer,
-	tree: treeReducer
+    modal: modalReducer,
```
Comment (GalGend) [inline client/src/redux/reducers.js:10]: No changes made, revert

File: client/src/scenes/Plans.css [hunk 9]
```diff
@@ -33,7 +33,12 @@ html body {
 }
 
 #plans-search-input {
-  margin-top: 1em;
+  /* margin-top: 1em; */
```
Comment (GalGend) [inline client/src/scenes/Plans.css:36]: remove comment

File: client/src/scenes/Plans.jsx [hunk 10]
```diff
@@ -32,38 +272,86 @@ class Plans extends Component {
         this.loadNextPage = this.loadNextPage.bind(this);
     }
 
+    handleSelectBox(event) {
+        this.setState({
+            selected: event.target.value,
+            state: true,
+        });
+        if (event.target.value === 'inputbox') {
```
Comment (GalGend) [inline client/src/scenes/Plans.jsx:280]: What is this?

File: client/src/shared/planCard/index.jsx [hunk 11]
```diff
@@ -18,57 +24,76 @@ import { useFavoritePlan } from '../../pages/Plan/hooks';
...
+                threshold: 0,
+            }
+        );
+
+        if (currentTagsRef) {
+            currentTagsRef.forEach((ref) => {
                 ref.current && observer.observe(ref.current);
-			});
-		}
+            });
```
Comment (GalGend) [inline client/src/shared/planCard/index.jsx:82]: No changes here, revert

File: client/src/shared/planCard/index.jsx [hunk 12]
```diff
@@ -199,7 +240,7 @@ const StatusDot = styled.div`
     flex-shrink: 0;
```
Comment (GalGend) [inline client/src/shared/planCard/index.jsx:240]: No CHANGES IN THIS FILE- REVERT

File: server/api/controller/plan.js [hunk 13]
```diff
@@ -1,58 +1,59 @@
...
-const Tag = require('../model/tag');
-const Log = require('../lib/log');
+const Controller = require("../controller/controller");
+const Plan = require("../model/plan");
+const GJV = require("geojson-validation");
+const { assign } = require("lodash");
+const Config = require("../lib/config");
+const { Knex } = require("../service/database");
+const Exception = require("../model/exception");
+const wkt = require("terraformer-wkt-parser");
```
Comment (GalGend) [inline server/api/controller/plan.js:8]: FIX LINTING

**Review by GalGend — COMMENTED** [review]
Please address all issues raised here

File: server/api/controller/shape.js (hunk 2 above)
Comment (harshitsan) [reply server/api/controller/shape.js]: yes

File: client/src/pages/Tree/common/Views/shareTree.jsx (hunk 3 above)
Comment (harshitsan) [reply client/src/pages/Tree/common/Views/shareTree.jsx]: updated back to original text

File: client/src/components/AutoCompleteParcleInput.jsx (hunk 4 above)
Comment (harshitsan) [reply client/src/components/AutoCompleteParcleInput.jsx]: removed

File: server/api/controller/shape.js [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+            msg: 'Please insert at least block number as a query param',
+        });
+    }
+
+    let sql;
+
+    // if both blockNum and parcelNum are passed as query param then this query will get selected and fetch from parcel_details table
+    // blockNum is being matched with gush_num(column name) and parcelNum with parcel(column name)
+    if (blockNum && parcelNum) {
+        sql = `SELECT id, ST_AsGeoJSON(centroid) as centroid, gush_num, parcel from parcel_details WHERE gush_num=${blockNum} and parcel=${parcelNum}`;
```
[the remark this answers is not in the document]
Comment (harshitsan) [reply server/api/controller/shape.js]: can you please verify again

[the remark this answers is not in the document]
Comment (harshitsan) [reply server/api/controller/shape.js]: there are two tables one with blocks only and one with parcels + block

File: server/api/controller/shape.js [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+    }
```
[the remark this answers is not in the document]
Comment (harshitsan) [reply server/api/controller/shape.js]: added

File: client/src/scenes/Plans.jsx [hunk 16]
```diff
@@ -149,12 +469,186 @@ class Plans extends Component {
...
+                    this.setState({
+                        searchPoint: searchLocation,
+                    });
+                    // this.loadPlans(this.state.pageNumber);
+
+                    this.loadPlans(this.state.pageNumber, searchLocation);
+                })
+                .catch((error) => this.setState({ error: '' }));
+        } else {
+            await api
```
[the remark this answers is not in the document]
Comment (harshitsan) [reply client/src/scenes/Plans.jsx]: fixed

File: client/src/scenes/Plans.jsx [hunk 17]
```diff
@@ -149,12 +469,186 @@ class Plans extends Component {
...
+                );
+        } else {
+            this.setState({
+                blockList: [],
+                loadingAutocompleteBlock: false,
+            });
+        }
+    }
+
+    async handleInputChangeParcle(text) {
```
[the remark this answers is not in the document]
Comment (harshitsan) [reply client/src/scenes/Plans.jsx]: updated

File: client/src/scenes/Plans.jsx [hunk 18]
```diff
@@ -149,12 +469,186 @@ class Plans extends Component {
...
+
+                    this.loadPlans(this.state.pageNumber, searchLocation);
+                    // Run the Above Query When Data is Comming From Other API
+                    // this.loadPlans(this.state.pageNumber);
+                })
+                .catch((error) => this.setState({ error: '' }));
+        }
+    }
+
+    async handleInputChangeBlock(text) {
```
[the remark this answers is not in the document]
Comment (harshitsan) [reply client/src/scenes/Plans.jsx]: updated

File: client/src/scenes/Plans.jsx [hunk 19]
```diff
@@ -149,12 +469,186 @@ class Plans extends Component {
...
+            noData: false,
+            pageNumber: 1,
+            searchPoint: false,
+            searchType: 'כתובת', //גוש חלקה
+            block: blockNum,
+            isDisable: false,
+        });
+    }
+
+    handleSubmitParcleDetails(parcleNum) {
```
[the remark this answers is not in the document]
Comment (harshitsan) [reply client/src/scenes/Plans.jsx]: updated

File: client/src/scenes/Plans.jsx [hunk 20]
```diff
@@ -167,43 +661,182 @@ class Plans extends Component {
...
+            plans,
+            error,
+            noData,
+            hasMore,
+            list,
+            loadingAutocomplete,
+            selected,
+            state,
+            block,
+            parcle,
```
[the remark this answers is not in the document]
Comment (harshitsan) [reply client/src/scenes/Plans.jsx]: updated

File: server/api/controller/plan.js [hunk 21]
```diff
@@ -63,135 +64,146 @@ class PlanController extends Controller {
 			// multiplied to get an approximate meters value
 			const spatialUnitFactor = Config.locationSearch.dbDistanceInMeters ? 1 : 111195;
 
-			q.columns.push(Knex.raw(`ST_Distance(geom, ST_GeomFromText("${polygon}",4326))*${spatialUnitFactor} as distance`));
+			//   console.log(geom);
+			q.columns.push(
+				Knex.raw(
+					`ST_Distance(ST_GeomFromText(ST_AsText(geom),4326), ST_GeomFromText("${polygon}",4326))*${spatialUnitFactor} as distance`
```
Comment (gruppin) [inline server/api/controller/plan.js:70]: @harshitsan  why did you change `geom` to `ST_GeomFromText(ST_AsText(geom)`?

Comment (harshitsan) [reply server/api/controller/plan.js:70]: Yes, Actually this is not required, This can be reverted to the master branch.
